src/antidote.erl

```diff
@@ -126,33 +126,59 @@ read_objects(Objects, TxId) ->
-                                lager:debug("Update failed. Reason : ~p",[Reason]),
-                                error
-                        end
-                end, Updates),
-    case lists:member(error, Results) of
-       true -> {error, update_failed}; %% TODO: Capture the reason for error
-       false -> ok
+    {_, _, CoordFsmPid} = TxId,
+    Operations = lists:map(fun(Update) ->
+                                {Key, Type, Bucket, Op} = case Update of
+                                    %% the following checks the format of the
+                                    %% operation for compatibility with
+                                    %% some systests that call updates with
+                                    %% {Operation, Params} (as a single parameter),
+                                    %% and Operation, Params (two separate parameters).
+                                    {{K, T, B}, O} ->
+                                        {K, T, B, O};
+                                    {{K, T, B}, O, P} ->
+                                        {K, T, B, {O, P}}
+                                end,
+                                case materializer:check_operations([{update, {{Key, Bucket}, Type, Op}}]) of
+                                    ok ->
+                                        {{Key, Bucket}, Type, Op};
+                                    {error, _Reason} ->
+                                        {error, type_check}
+                                end
+                            end, Updates),
+    case lists:member({error, type_check}, Operations) of
+        true ->
+            {error, type_check};
+        false ->
+            %%            lager:info("gonna start multiple updates: ~p", [Operations]),
+            case gen_fsm:sync_send_event(CoordFsmPid, {update_objects, Operations}, ?OP_TIMEOUT) of
+                ok ->
+                    ok;
+                {aborted, TxId} ->
+                    {aborted, TxId};
+                {error, Reason} ->
+                    {error, Reason}
+            end
     end.
 
 %% For static transactions: bulk updates and bulk reads
 -spec update_objects(snapshot_time() | ignore , term(), [{bound_object(), op_name(), op_param()}]) ->
                             {ok, snapshot_time()} | {error, reason()}.
 update_objects(Clock, Properties, Updates) ->
     update_objects(Clock, Properties, Updates, false).
-
+update_objects(_Clock, _Properties, [], _StayAlive) ->
+    {ok, vectorclock:new()};
 update_objects(Clock, _Properties, Updates, StayAlive) ->
-    Operations = lists:map(
-                   fun({{Key, Type, Bucket}, Op, OpParam}) ->
-                           {update, {{Key, Bucket}, Type, {Op,OpParam}}}
-                   end,
+    Operations = lists:map(fun(Update) ->
+                                {Key, Type, Bucket, Op} =
+                                    case Update of
+                                    {{K, T, B}, O} ->
```

Comment:
If it is really, really needed, you should refactor it into a separate function.